sql: TestRaceWithBackfill failed

[(sql).TestRaceWithBackfill failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2359994&tab=buildLog) on [master@5ca5b10632167d0ccc726403693e22d26e07e180](https://github.com/cockroachdb/cockroach/commits/5ca5b10632167d0ccc726403693e22d26e07e180):

```
=== RUN   TestRaceWithBackfill
*
* WARNING: disk slowness detected: unable to sync log files within 10s
*
*
* FATAL: disk stall detected: unable to sync log files within 1m0s
*


ERROR: exit status 1

1 runs completed, 1 failures, over 27m4s
context canceled


Failed to find issue assignee: 
old friend vivekmenezes is friendless
```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestRaceWithBackfill PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRaceWithBackfill.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestRaceWithBackfill failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2363610&tab=buildLog) on [master@05826e3a748d3bf7634130d49667627e687d5875](https://github.com/cockroachdb/cockroach/commits/05826e3a748d3bf7634130d49667627e687d5875):

Fatal error:
```
panic: interface conversion: tree.Datum is tree.dNull, not *tree.DBytes [recovered]
	panic: interface conversion: tree.Datum is tree.dNull, not *tree.DBytes

```

Stack:
```
goroutine 399123 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc09e892500, 0x885b7e0, 0xc0bb563f40)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:207 +0x154
panic(0x6320ea0, 0xc08f753a40)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).Update.func1(0x885b8a0, 0xc067ffde30, 0xc064fa1dd0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:132 +0x216e
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1(0x885b8a0, 0xc067ffde30, 0xc064fa1dd0, 0xc07504c900, 0x89d7d60)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:707 +0x5f
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc064fa1dd0, 0x885b8a0, 0xc067ffde30, 0xc09438ece0, 0xc064fa1dd0, 0x424)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:811 +0xea
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0xc08e83d880, 0x885b8a0, 0xc067ffde30, 0xc09438ee10, 0x0, 0x6c3a50)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:706 +0x11d
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).runInTxn(0xc0758d43c0, 0x885b8a0, 0xc067ffde30, 0xc09438ee10, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:693 +0x181
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).Update(0xc0758d43c0, 0x885b8a0, 0xc067ffde30, 0xc09438ee88, 0x12ef5ef, 0x885b8a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:108 +0x10c
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).reverted(0xc0758d43c0, 0x885b8a0, 0xc067ffde30, 0x87c0860, 0xc073c9d620, 0x0, 0xc09438f418, 0x4)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:519 +0xb3
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc0c2212c60, 0x885b8a0, 0xc067ffde30, 0x6887cc0, 0xc08a3e0000, 0x8806b20, 0xc030d58110, 0xc0962670e0, 0xc0758d43c0, 0x689ca7e, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1138 +0x2b74
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc0c2212c60, 0x885b8a0, 0xc067ffde30, 0x6887cc0, 0xc08a3e0000, 0x8806b20, 0xc030d58110, 0xc0962670e0, 0xc0758d43c0, 0x6897bf2, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1108 +0x1ea6
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc0c2212c60, 0x885b8a0, 0xc067ffde30, 0x8806b20, 0xc030d58110, 0xc0962670e0, 0xc0855fc960, 0xc0758d43c0, 0x6897bf2, 0x7, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:245 +0x46b
github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start.func2(0x885b7e0, 0xc0bb563f40)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:853 +0x2ca
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc09e892500, 0x885b7e0, 0xc0bb563f40, 0xc08f8b5f60, 0x1e, 0x89b3900, 0xc095fce700, 0xc072cf7a60)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347 +0x163
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:342 +0x14b

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestRaceWithBackfill

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestRaceWithBackfill PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRaceWithBackfill.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestRaceWithBackfill failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2366543&tab=buildLog) on [master@00d70ee5684f1d50000e6ce215f2af960d35602b](https://github.com/cockroachdb/cockroach/commits/00d70ee5684f1d50000e6ce215f2af960d35602b):

Fatal error:
```
panic: interface conversion: tree.Datum is tree.dNull, not *tree.DBytes [recovered]
	panic: interface conversion: tree.Datum is tree.dNull, not *tree.DBytes

```

Stack:
```
goroutine 392654 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc072012780, 0x897b200, 0xc07f1b4680)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:207 +0x154
panic(0x63d72c0, 0xc01efc0e70)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).Update.func1(0x897b2c0, 0xc05d7fd140, 0xc089a11050, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:132 +0x216e
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1(0x897b2c0, 0xc05d7fd140, 0xc089a11050, 0xc082c12900, 0x8b085a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:707 +0x5f
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc089a11050, 0x897b2c0, 0xc05d7fd140, 0xc0a94cece8, 0xc089a11050, 0xc0505d13e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:811 +0xea
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0xc05278e200, 0x897b2c0, 0xc05d7fd140, 0xc0a94cee18, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:706 +0x11d
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).runInTxn(0xc0158ee780, 0x897b2c0, 0xc05d7fd140, 0xc0a94cee18, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:693 +0x181
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).Update(0xc0158ee780, 0x897b2c0, 0xc05d7fd140, 0xc0a94cee90, 0x897b2c0, 0xc05d7fd140)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:108 +0x10c
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).succeeded(0xc0158ee780, 0x897b2c0, 0xc05d7fd140, 0x0, 0x0, 0x34)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:594 +0x9f
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc059c8a840, 0x897b2c0, 0xc05d7fd140, 0x69a1300, 0xc095c93600, 0x89219c0, 0xc01dce8430, 0xc058154ae0, 0xc0158ee780, 0x69b64db, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1128 +0x2817
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc059c8a840, 0x897b2c0, 0xc05d7fd140, 0x69a1300, 0xc095c93600, 0x89219c0, 0xc01dce8430, 0xc058154ae0, 0xc0158ee780, 0x69b1232, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1082 +0x2157
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc059c8a840, 0x897b2c0, 0xc05d7fd140, 0x89219c0, 0xc01dce8430, 0xc058154ae0, 0xc06e7883c0, 0xc0158ee780, 0x69b1232, 0x7, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:245 +0x46b
github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start.func2(0x897b200, 0xc07f1b4680)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:853 +0x2ca
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc072012780, 0x897b200, 0xc07f1b4680, 0xc030fb83c0, 0x1e, 0x8ae3f40, 0xc05577afe0, 0xc0249e1700)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347 +0x163
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:342 +0x14b

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestRaceWithBackfill
*
* WARNING: disk slowness detected: unable to sync log files within 10s
*

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestRaceWithBackfill PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRaceWithBackfill.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestRaceWithBackfill failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2369314&tab=buildLog) on [master@bb6c84f14d8888a9e58826d5e72cc404ebfd0fde](https://github.com/cockroachdb/cockroach/commits/bb6c84f14d8888a9e58826d5e72cc404ebfd0fde):

Fatal error:
```
panic: interface conversion: tree.Datum is tree.dNull, not *tree.DBytes [recovered]
	panic: interface conversion: tree.Datum is tree.dNull, not *tree.DBytes

```

Stack:
```
goroutine 400096 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0c2e34820, 0x898dd20, 0xc082cad2c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:207 +0x154
panic(0x63e1ac0, 0xc07a0b1d40)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).Update.func1(0x898dde0, 0xc08c403410, 0xc04c0d8900, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:132 +0x216e
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1(0x898dde0, 0xc08c403410, 0xc04c0d8900, 0xc084a22900, 0x8b1b220)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:707 +0x5f
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc04c0d8900, 0x898dde0, 0xc08c403410, 0xc0a4422ce8, 0xc04c0d8900, 0xc0abf0b500)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:811 +0xea
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0xc0977a6e80, 0x898dde0, 0xc08c403410, 0xc0a4422e18, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:706 +0x11d
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).runInTxn(0xc081ed8dc0, 0x898dde0, 0xc08c403410, 0xc0a4422e18, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:693 +0x181
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).Update(0xc081ed8dc0, 0x898dde0, 0xc08c403410, 0xc0a4422e90, 0x898dde0, 0xc08c403410)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:108 +0x10c
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).succeeded(0xc081ed8dc0, 0x898dde0, 0xc08c403410, 0x0, 0x0, 0x34)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:594 +0x9f
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc0c23838c0, 0x898dde0, 0xc08c403410, 0x69ad760, 0xc0b8873600, 0x8934320, 0xc02e389b70, 0xc09e4b9860, 0xc081ed8dc0, 0x69c2981, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1128 +0x2817
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc0c23838c0, 0x898dde0, 0xc08c403410, 0x69ad760, 0xc0b8873600, 0x8934320, 0xc02e389b70, 0xc09e4b9860, 0xc081ed8dc0, 0x69bd6d8, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1082 +0x2157
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc0c23838c0, 0x898dde0, 0xc08c403410, 0x8934320, 0xc02e389b70, 0xc09e4b9860, 0xc07eb8b3e0, 0xc081ed8dc0, 0x69bd6d8, 0x7, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:245 +0x46b
github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start.func2(0x898dd20, 0xc082cad2c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:853 +0x2ca
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc0c2e34820, 0x898dd20, 0xc082cad2c0, 0xc0541403a0, 0x1e, 0x8af6bc0, 0xc09904c1e0, 0xc075988e80)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347 +0x163
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:342 +0x14b

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestRaceWithBackfill

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestRaceWithBackfill PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRaceWithBackfill.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestRaceWithBackfill failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2416848&tab=buildLog) on [master@25edeb8d6c2cefca0c206f86553bc75569f0f2d9](https://github.com/cockroachdb/cockroach/commits/25edeb8d6c2cefca0c206f86553bc75569f0f2d9):

Fatal error:
```
panic: interface conversion: tree.Datum is tree.dNull, not *tree.DBytes [recovered]
	panic: interface conversion: tree.Datum is tree.dNull, not *tree.DBytes

```

Stack:
```
goroutine 390607 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0cb4ab220, 0x838c760, 0xc030f17340)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:206 +0x154
panic(0x5d9eb00, 0xc08df89f20)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).Update.func1(0x838c820, 0xc077e25440, 0xc010078a20, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:132 +0x21ad
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1(0x838c820, 0xc077e25440, 0xc010078a20, 0xc086cf0480, 0x85184e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:707 +0x5f
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc010078a20, 0x838c820, 0xc077e25440, 0xc0c48d0cf8, 0xc010078a20, 0xc09c3ed740)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:811 +0xea
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0xc0c4f7d9d0, 0x838c820, 0xc077e25440, 0xc0c48d0e28, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:706 +0x11d
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).runInTxn(0xc077d09540, 0x838c820, 0xc077e25440, 0xc0c48d0e28, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:693 +0x181
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).Update(0xc077d09540, 0x838c820, 0xc077e25440, 0xc0c48d0ea0, 0x838c820, 0xc077e25440)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:108 +0x10c
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).succeeded(0xc077d09540, 0x838c820, 0xc077e25440, 0x0, 0x0, 0x34)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:594 +0x9f
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc059e7a000, 0x838c820, 0xc077e25440, 0x6006380, 0xc07c03d8f8, 0x8332a60, 0xc0b98da960, 0xc086890180, 0xc077d09540, 0x63863e4, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1138 +0x2817
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc059e7a000, 0x838c820, 0xc077e25440, 0x6006380, 0xc07c03d8f8, 0x8332a60, 0xc0b98da960, 0xc086890180, 0xc077d09540, 0x63810a9, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1092 +0x2157
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc059e7a000, 0x838c820, 0xc077e25440, 0x8332a60, 0xc0b98da960, 0xc086890180, 0xc09e123bc0, 0xc077d09540, 0x63810a9, 0x7, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:245 +0x461
github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start.func2(0x838c760, 0xc030f17340)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:853 +0x2ca
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc0cb4ab220, 0x838c760, 0xc030f17340, 0xc057caba40, 0x1e, 0xc01c4607e0, 0xc04f6f0780)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:346 +0x156
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:341 +0x135

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestRaceWithBackfill

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestRaceWithBackfill PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRaceWithBackfill.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestRaceWithBackfill failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2430622&tab=buildLog) on [master@f2b53d1080f1ccc503263dd178ee31b4878d2360](https://github.com/cockroachdb/cockroach/commits/f2b53d1080f1ccc503263dd178ee31b4878d2360):

Fatal error:
```
panic: interface conversion: tree.Datum is tree.dNull, not *tree.DBytes [recovered]
	panic: interface conversion: tree.Datum is tree.dNull, not *tree.DBytes

```

Stack:
```
goroutine 392748 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0901a12c0, 0x85ebd80, 0xc069f28cc0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:206 +0x154
panic(0x5f92620, 0xc07cad9a40)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).Update.func1(0x85ebe40, 0xc0a3fb2720, 0xc05330fc20, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:132 +0x21ad
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1(0x85ebe40, 0xc0a3fb2720, 0xc05330fc20, 0xc07bd52480, 0x8778240)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:707 +0x5f
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc05330fc20, 0x85ebe40, 0xc0a3fb2720, 0xc0a6284cf8, 0xc05330fc20, 0xc087a3d140)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:811 +0xea
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0xc01d38ec40, 0x85ebe40, 0xc0a3fb2720, 0xc0a6284e28, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:706 +0x11d
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).runInTxn(0xc08eedc640, 0x85ebe40, 0xc0a3fb2720, 0xc0a6284e28, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:693 +0x181
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).Update(0xc08eedc640, 0x85ebe40, 0xc0a3fb2720, 0xc0a6284ea0, 0x85ebe40, 0xc0a3fb2720)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/update.go:108 +0x10c
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).succeeded(0xc08eedc640, 0x85ebe40, 0xc0a3fb2720, 0x0, 0x0, 0x34)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:594 +0x9f
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc0a2c85a40, 0x85ebe40, 0xc0a3fb2720, 0x61fa200, 0xc08464f140, 0x8591a00, 0xc077d73140, 0xc0c4e3f980, 0xc08eedc640, 0x657c4c2, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1138 +0x2817
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc0a2c85a40, 0x85ebe40, 0xc0a3fb2720, 0x61fa200, 0xc08464f140, 0x8591a00, 0xc077d73140, 0xc0c4e3f980, 0xc08eedc640, 0x657715a, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1092 +0x2157
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc0a2c85a40, 0x85ebe40, 0xc0a3fb2720, 0x8591a00, 0xc077d73140, 0xc0c4e3f980, 0xc081056360, 0xc08eedc640, 0x657715a, 0x7, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:245 +0x461
github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start.func2(0x85ebd80, 0xc069f28cc0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:853 +0x2ca
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc0901a12c0, 0x85ebd80, 0xc069f28cc0, 0xc0763d5520, 0x1e, 0xc03114b100, 0xc03e5d5000)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:346 +0x156
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:341 +0x135

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestRaceWithBackfill

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestRaceWithBackfill PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRaceWithBackfill.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

The failures that say the following seem real
```
panic: interface conversion: tree.Datum is tree.dNull, not *tree.DBytes [recovered]
	panic: interface conversion: tree.Datum is tree.dNull, not *tree.DBytes
```

The ones complaining about disk slowness are not specific to this test.

--- a/pkg/sql/colexec/colbuilder/execplan.go
+++ b/pkg/sql/colexec/colbuilder/execplan.go
@@ -762,10 +762,13 @@ func NewColOperator(
 			cFetcherMemAcc := result.createUnlimitedMemAccount(
 				ctx, flowCtx, "cfetcher" /* opName */, spec.ProcessorID,
 			)
+			kvFetcherMemAcc := result.createUnlimitedMemAccount(
+				ctx, flowCtx, "kvfetcher" /* opName */, spec.ProcessorID,
+			)
 			estimatedRowCount := spec.EstimatedRowCount
 			scanOp, err := colfetcher.NewColBatchScan(
-				ctx, colmem.NewAllocator(ctx, cFetcherMemAcc, factory), flowCtx,
-				evalCtx, core.TableReader, post, estimatedRowCount,
+				ctx, colmem.NewAllocator(ctx, cFetcherMemAcc, factory), kvFetcherMemAcc,
+				flowCtx, evalCtx, core.TableReader, post, estimatedRowCount,
 			)
 			if err != nil {
 				return r, err
@@ -786,11 +789,14 @@ func NewColOperator(
 			cFetcherMemAcc := result.createUnlimitedMemAccount(
 				ctx, flowCtx, "cfetcher" /* opName */, spec.ProcessorID,
 			)
+			kvFetcherMemAcc := result.createUnlimitedMemAccount(
+				ctx, flowCtx, "kvfetcher" /* opName */, spec.ProcessorID,
+			)
 			semaCtx := flowCtx.TypeResolverFactory.NewSemaContext(evalCtx.Txn)
 			inputTypes := make([]*types.T, len(spec.Input[0].ColumnTypes))
 			copy(inputTypes, spec.Input[0].ColumnTypes)
 			indexJoinOp, err := colfetcher.NewColIndexJoin(
-				ctx, streamingAllocator, colmem.NewAllocator(ctx, cFetcherMemAcc, factory),
+				ctx, streamingAllocator, colmem.NewAllocator(ctx, cFetcherMemAcc, factory), kvFetcherMemAcc,
 				flowCtx, evalCtx, semaCtx, inputs[0].Root, core.JoinReader, post, inputTypes)
 			if err != nil {
 				return r, err

--- a/pkg/sql/colfetcher/BUILD.bazel
+++ b/pkg/sql/colfetcher/BUILD.bazel
@@ -40,6 +40,7 @@ go_library(
         "//pkg/util/encoding",
         "//pkg/util/hlc",
         "//pkg/util/log",
+        "//pkg/util/mon",
         "//pkg/util/syncutil",
         "//pkg/util/tracing",
         "@com_github_cockroachdb_apd_v2//:apd",

--- a/pkg/sql/colfetcher/cfetcher.go
+++ b/pkg/sql/colfetcher/cfetcher.go
@@ -43,6 +43,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/errors"
 )
 
@@ -337,6 +338,10 @@ type cFetcher struct {
 	accountingHelper colmem.SetAccountingHelper
 	memoryLimit      int64
 
+	// kvFetcherMemAcc is a memory account that will be used by the underlying
+	// KV fetcher.
+	kvFetcherMemAcc *mon.BoundAccount
+
 	// maxCapacity if non-zero indicates the target capacity of the output
 	// batch. It is set when at the row finalization we realize that the output
 	// batch has exceeded the memory limit.
@@ -395,6 +400,7 @@ func (rf *cFetcher) resetBatch() {
 func (rf *cFetcher) Init(
 	codec keys.SQLCodec,
 	allocator *colmem.Allocator,
+	kvFetcherMemAcc *mon.BoundAccount,
 	memoryLimit int64,
 	reverse bool,
 	lockStrength descpb.ScanLockingStrength,
@@ -403,6 +409,7 @@ func (rf *cFetcher) Init(
 	tableArgs row.FetcherTableArgs,
 	traceKV bool,
 ) error {
+	rf.kvFetcherMemAcc = kvFetcherMemAcc
 	rf.memoryLimit = memoryLimit
 	rf.reverse = reverse
 	rf.lockStrength = lockStrength
@@ -711,7 +718,7 @@ func (rf *cFetcher) StartScan(
 		rf.lockStrength,
 		rf.lockWaitPolicy,
 		rf.lockTimeout,
-		rf.accountingHelper.Allocator.GetMonitor(),
+		rf.kvFetcherMemAcc,
 		forceProductionKVBatchSize,
 	)
 	if err != nil {
@@ -1620,6 +1627,7 @@ type cFetcherArgs struct {
 func initCFetcher(
 	flowCtx *execinfra.FlowCtx,
 	allocator *colmem.Allocator,
+	kvFetcherMemAcc *mon.BoundAccount,
 	desc catalog.TableDescriptor,
 	index catalog.Index,
 	neededCols util.FastIntSet,
@@ -1641,7 +1649,7 @@ func initCFetcher(
 	tableArgs.InitCols(desc, args.visibility, args.hasSystemColumns, invertedCol)
 
 	if err := fetcher.Init(
-		flowCtx.Codec(), allocator, args.memoryLimit, args.reverse, args.lockingStrength,
+		flowCtx.Codec(), allocator, kvFetcherMemAcc, args.memoryLimit, args.reverse, args.lockingStrength,
 		args.lockingWaitPolicy, flowCtx.EvalCtx.SessionData().LockTimeout, tableArgs, flowCtx.TraceKV,
 	); err != nil {
 		return nil, err

--- a/pkg/sql/colfetcher/colbatch_scan.go
+++ b/pkg/sql/colfetcher/colbatch_scan.go
@@ -28,6 +28,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/cockroachdb/errors"
@@ -184,6 +185,7 @@ var colBatchScanPool = sync.Pool{
 func NewColBatchScan(
 	ctx context.Context,
 	allocator *colmem.Allocator,
+	kvFetcherMemAcc *mon.BoundAccount,
 	flowCtx *execinfra.FlowCtx,
 	evalCtx *tree.EvalContext,
 	spec *execinfrapb.TableReaderSpec,
@@ -218,7 +220,7 @@ func NewColBatchScan(
 	}
 
 	fetcher, err := initCFetcher(
-		flowCtx, allocator, table, table.ActiveIndexes()[spec.IndexIdx],
+		flowCtx, allocator, kvFetcherMemAcc, table, table.ActiveIndexes()[spec.IndexIdx],
 		neededColumns, columnIdxMap, invertedColumn,
 		cFetcherArgs{
 			visibility:        spec.Visibility,

--- a/pkg/sql/colfetcher/index_join.go
+++ b/pkg/sql/colfetcher/index_join.go
@@ -31,6 +31,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/cockroachdb/errors"
@@ -385,6 +386,7 @@ func NewColIndexJoin(
 	ctx context.Context,
 	allocator *colmem.Allocator,
 	fetcherAllocator *colmem.Allocator,
+	kvFetcherMemAcc *mon.BoundAccount,
 	flowCtx *execinfra.FlowCtx,
 	evalCtx *tree.EvalContext,
 	semaCtx *tree.SemaContext,
@@ -458,7 +460,7 @@ func NewColIndexJoin(
 	}
 
 	fetcher, err := initCFetcher(
-		flowCtx, fetcherAllocator, table, index, neededColumns, columnIdxMap, nil, /* invertedColumn */
+		flowCtx, fetcherAllocator, kvFetcherMemAcc, table, index, neededColumns, columnIdxMap, nil, /* invertedColumn */
 		cFetcherArgs{
 			visibility:        spec.Visibility,
 			lockingStrength:   spec.LockingStrength,

--- a/pkg/sql/colmem/allocator.go
+++ b/pkg/sql/colmem/allocator.go
@@ -119,12 +119,6 @@ func NewAllocator(
 	}
 }
 
-// GetMonitor returns the bytes monitor which the allocator's memory account is
-// bound to.
-func (a *Allocator) GetMonitor() *mon.BytesMonitor {
-	return a.acc.Monitor()
-}
-
 // NewMemBatchWithFixedCapacity allocates a new in-memory coldata.Batch with the
 // given vector capacity.
 // Note: consider whether you want the dynamic batch size behavior (in which

--- a/pkg/sql/row/fetcher.go
+++ b/pkg/sql/row/fetcher.go
@@ -289,8 +289,9 @@ type Fetcher struct {
 	// Buffered allocation of decoded datums.
 	alloc *rowenc.DatumAlloc
 
-	// Memory monitor for the bytes fetched by this fetcher.
-	mon *mon.BytesMonitor
+	// Memory monitor and memory account for the bytes fetched by this fetcher.
+	mon             *mon.BytesMonitor
+	kvFetcherMemAcc *mon.BoundAccount
 }
 
 // Reset resets this Fetcher, preserving the memory capacity that was used
@@ -309,6 +310,7 @@ func (rf *Fetcher) Close(ctx context.Context) {
 		rf.kvFetcher.Close(ctx)
 	}
 	if rf.mon != nil {
+		rf.kvFetcherMemAcc.Close(ctx)
 		rf.mon.Stop(ctx)
 	}
 }
@@ -343,6 +345,8 @@ func (rf *Fetcher) Init(
 	if memMonitor != nil {
 		rf.mon = mon.NewMonitorInheritWithLimit("fetcher-mem", 0 /* limit */, memMonitor)
 		rf.mon.Start(ctx, memMonitor, mon.BoundAccount{})
+		memAcc := rf.mon.MakeBoundAccount()
+		rf.kvFetcherMemAcc = &memAcc
 	}
 
 	// We must always decode the index key if we need to distinguish between
@@ -605,7 +609,7 @@ func (rf *Fetcher) StartScan(
 		rf.lockStrength,
 		rf.lockWaitPolicy,
 		rf.lockTimeout,
-		rf.mon,
+		rf.kvFetcherMemAcc,
 		forceProductionKVBatchSize,
 		txn.AdmissionHeader(),
 		txn.DB().SQLKVResponseAdmissionQ,
@@ -706,7 +710,7 @@ func (rf *Fetcher) StartInconsistentScan(
 		rf.lockStrength,
 		rf.lockWaitPolicy,
 		rf.lockTimeout,
-		rf.mon,
+		rf.kvFetcherMemAcc,
 		forceProductionKVBatchSize,
 		txn.AdmissionHeader(),
 		txn.DB().SQLKVResponseAdmissionQ,

--- a/pkg/sql/row/kv_batch_fetcher.go
+++ b/pkg/sql/row/kv_batch_fetcher.go
@@ -107,7 +107,7 @@ type txnKVFetcher struct {
 	responses        []roachpb.ResponseUnion
 	remainingBatches [][]byte
 
-	acc mon.BoundAccount
+	acc *mon.BoundAccount
 	// spansAccountedFor and batchResponseAccountedFor track the number of bytes
 	// that we've already registered with acc in regards to spans and the batch
 	// response, respectively.
@@ -240,11 +240,15 @@ func makeKVBatchFetcherDefaultSendFunc(txn *kv.Txn) sendFunc {
 // 0) to ProductionKVBatchSize.
 //
 // The fetcher takes ownership of the spans slice - it can modify the slice and
-// will perform the memory accounting accordingly (if mon is non-nil). The
+// will perform the memory accounting accordingly (if acc is non-nil). The
 // caller can only reuse the spans slice after the fetcher has been closed, and
 // if the caller does, it becomes responsible for the memory accounting.
 //
 // Batch limits can only be used if the spans are ordered.
+//
+// The passed-in memory account is owned by the fetcher throughout its lifetime
+// but is **not** closed - it is the caller's responsibility to close acc if it
+// is non-nil.
 func makeKVBatchFetcher(
 	ctx context.Context,
 	sendFn sendFunc,
@@ -255,7 +259,7 @@ func makeKVBatchFetcher(
 	lockStrength descpb.ScanLockingStrength,
 	lockWaitPolicy descpb.ScanLockingWaitPolicy,
 	lockTimeout time.Duration,
-	mon *mon.BytesMonitor,
+	acc *mon.BoundAccount,
 	forceProductionKVBatchSize bool,
 	requestAdmissionHeader roachpb.AdmissionHeader,
 	responseAdmissionQ *admission.WorkQueue,
@@ -316,14 +320,14 @@ func makeKVBatchFetcher(
 		lockStrength:               lockStrength,
 		lockWaitPolicy:             lockWaitPolicy,
 		lockTimeout:                lockTimeout,
-		acc:                        mon.MakeBoundAccount(),
+		acc:                        acc,
 		forceProductionKVBatchSize: forceProductionKVBatchSize,
 		requestAdmissionHeader:     requestAdmissionHeader,
 		responseAdmissionQ:         responseAdmissionQ,
 	}
 
 	// Account for the memory of the spans that we're taking the ownership of.
-	if f.acc.Monitor() != nil {
+	if f.acc != nil {
 		f.spansAccountedFor = spans.MemUsage()
 		if err := f.acc.Grow(ctx, f.spansAccountedFor); err != nil {
 			return txnKVFetcher{}, err
@@ -437,7 +441,7 @@ func (f *txnKVFetcher) fetch(ctx context.Context) error {
 		log.VEventf(ctx, 2, "Scan %s", buf.String())
 	}
 
-	monitoring := f.acc.Monitor() != nil
+	monitoring := f.acc != nil
 
 	const tokenFetchAllocation = 1 << 10
 	if !monitoring || f.batchResponseAccountedFor < tokenFetchAllocation {
@@ -606,7 +610,7 @@ func (f *txnKVFetcher) nextBatch(
 		}
 		// We have some resume spans.
 		f.spans = f.spansScratch[:f.newFetchSpansIdx]
-		if f.acc.Monitor() != nil {
+		if f.acc != nil {
 			used := f.acc.Used()
 			delta := f.spans.MemUsage() - f.spansAccountedFor
 			if err := f.acc.Resize(ctx, used, used+delta); err != nil {
@@ -629,5 +633,5 @@ func (f *txnKVFetcher) close(ctx context.Context) {
 	f.remainingBatches = nil
 	f.spans = nil
 	f.spansScratch = nil
-	f.acc.Close(ctx)
+	f.acc.Clear(ctx)
 }

--- a/pkg/sql/row/kv_fetcher.go
+++ b/pkg/sql/row/kv_fetcher.go
@@ -44,10 +44,10 @@ type KVFetcher struct {
 }
 
 // NewKVFetcher creates a new KVFetcher.
-// If mon is non-nil, this fetcher will track its fetches and must be Closed.
+// If acc is non-nil, this fetcher will track its fetches and must be Closed.
 //
 // The fetcher takes ownership of the spans slice - it can modify the slice and
-// will perform the memory accounting accordingly (if mon is non-nil). The
+// will perform the memory accounting accordingly (if acc is non-nil). The
 // caller can only reuse the spans slice after the fetcher has been closed, and
 // if the caller does, it becomes responsible for the memory accounting.
 func NewKVFetcher(
@@ -61,7 +61,7 @@ func NewKVFetcher(
 	lockStrength descpb.ScanLockingStrength,
 	lockWaitPolicy descpb.ScanLockingWaitPolicy,
 	lockTimeout time.Duration,
-	mon *mon.BytesMonitor,
+	acc *mon.BoundAccount,
 	forceProductionKVBatchSize bool,
 ) (*KVFetcher, error) {
 	var sendFn sendFunc
@@ -100,7 +100,7 @@ func NewKVFetcher(
 		lockStrength,
 		lockWaitPolicy,
 		lockTimeout,
-		mon,
+		acc,
 		forceProductionKVBatchSize,
 		txn.AdmissionHeader(),
 		txn.DB().SQLKVResponseAdmissionQ,

